src/common/rsvc.c

```diff
@@ -104,8 +104,22 @@ rsvc_client_process_error(struct rsvc_client *client, int rc,
 {
 	int leader_index = client->sc_leader_index;
 
-	if (client->sc_leader_known && client->sc_leader_aliveness > 0 &&
-	    ep->ep_rank == client->sc_ranks->rl_ranks[leader_index]) {
+	if (rc == -DER_NONEXIST) {
+		int rc_tmp;
+
+		rc_tmp = daos_rank_list_del(client->sc_ranks, ep->ep_rank);
+		if (rc_tmp == -DER_NOMEM)
+			client->sc_ranks = NULL;
```

Comment:
A NULL client->sc_ranks looks pretty troublesome to deal with, as we assume it is always not NULL in the rest of this file. d_rank_list_del (i.e., the d_rank_list_realloc call) should not free the rank list---crazy, IMHO. Could we do the following instead, keeping the invariant that sc_ranks != NULL:


    if (d_rank_list_find(client->sc_ranks, ep->ep_rank, &i)) {
        remove the rank using memmove;
        client->sc_ranks->rl_nr--;

        client->sc_leader_known = false;
        client->sc_leader_aliveness = 0;
        client->sc_leader_term = -1;
        client->sc_leader_index = -1;

        client->sc_next %= client->sc_ranks->rl_nr;
    }

Trying to keeping the leader hint after removing a rank seems complex, so I just clear the leader hint here.

Reply:
probably d_rank_list_del should provide a non-shrink version.  It is a bit strange to see rank_list_del return NOMEM.